lib/mayaUsd/ufe/UsdStageMap.cpp

```diff
@@ -85,32 +110,53 @@ void UsdStageMap::addItem(const Ufe::Path& path, UsdStageWeakPtr stage)
         return;
     }
 
-    // Convert the tail of the UFE path to an MObjectHandle.
-    auto proxyShape = proxyShapeHandle(path);
+    // Convert the UFE path to an MObjectHandle.
+    auto proxyShape = nameLookup(path);
     if (!proxyShape.isValid()) {
         return;
     }
 
-    // Could get the stage from the proxy shape object in the stage() method,
-    // but since it's given here, simply store it.
-    fObjectToStage[proxyShape] = stage;
+    // We've just done a name-based lookup of the proxy shape, so the stage
+    // cannot be null.
+    auto stage = objToStage(proxyShape.object());
+    TF_AXIOM(stage);
+
+    fPathToObject[path] = proxyShape;
     fStageToObject[stage] = proxyShape;
 }
 
-UsdStageWeakPtr UsdStageMap::stage(const Ufe::Path& path)
+UsdStageWeakPtr UsdStageMap::stage(const Ufe::Path& path) { return objToStage(proxyShape(path)); }
+
+MObject UsdStageMap::proxyShape(const Ufe::Path& path)
 {
+    const auto& singleSegmentPath
+        = nbPathSegments(path) == 1 ? path : Ufe::Path(path.getSegments()[0]);
+
+    auto iter = fPathToObject.find(singleSegmentPath);
+    if (iter != std::end(fPathToObject)) {
+        return iter->second.object();
+    }
+
+    // Did not find object in cache, either because the cache is stale, or
```

Comment:
If the lookup fails, the cache might be stale.  Rebuild and try again.

Reply:
This is surprising. We should have proper invalidation of the cache on data model changes. Maybe this refactoring changed something and now we rebuild the cache too early (i.e. cause compute during dirty propagation) - this should be considered as a bug.

 Did you happen to find a case where this explicit setDirty was needed?

Reply:
The only invalidation on data model change we do is for file new, which is done here:
https://github.com/Autodesk/maya-usd/blob/9a9642151be2d5d50e647abc3a5a8bc85c48b2b2/lib/mayaUsd/ufe/StagesSubject.cpp#L195
The version of the stage cache I am changing does not do scene observation for reparenting and renaming --- and the reason it doesn't is that it always does name-based lookup.  Once it does that, it obtains an MObjectHandle, and uses that as a key to index into a map to retrieve the stage.  However, this "caching" is not very effective: it is trivial to retrieve the stage once you have the MObject.

The version of the cache I am proposing does not observe the data model for rename and reparent either, and the rationale is explained here:
https://github.com/Autodesk/maya-usd/blob/538b360ba7f6be6df30ab64cb189750e493d92ee/lib/mayaUsd/ufe/UsdStageMap.h#L47
Essentially, we completely avoid order of notification problems with the update on demand scheme I am proposing, and it's much simpler as well.

Reply:
The name may be here confusing, but we actually observe the data model and invalidate when stage gets invalidated during dirty propagation (or parallel evaluation) - https://github.com/Autodesk/maya-usd/blob/9a9642151be2d5d50e647abc3a5a8bc85c48b2b2/lib/mayaUsd/ufe/StagesSubject.cpp#L450

Reply:
Ah, thanks for the tip, yes the naming is indeed a bit confusing.  Does this handle renaming and reparenting?  My intuition is that it does not, because this is purely a DG invalidation mechanism, in which case I believe my stale cache handling is still required.